test/integration/device-state.spec.ts

```diff
@@ -616,19 +627,183 @@ describe('device-state', () => {
+				await expect(deviceState.shutdown())
+					.to.eventually.be.rejectedWith(errMsg)
+					.and.be.an.instanceOf(UpdatesLockedError);
+
+				await testfs.restore();
+			}
+		});
 
-		await expect(deviceState.shutdown({ reboot: true }))
-			.to.eventually.be.rejectedWith(testErrMsg)
-			.and.be.an.instanceOf(UpdatesLockedError);
-		await expect(deviceState.shutdown())
-			.to.eventually.be.rejectedWith(testErrMsg)
-			.and.be.an.instanceOf(UpdatesLockedError);
+		it('prevents reboot or shutdown when user service locks are present', async () => {
+			// Set up a user service lock
+			sinon.stub(applicationManager, 'getCurrentApps').resolves(
+				createApps(
+					{
+						services: [
+							await createService({
+								running: true,
+								appId: 123,
+								serviceName: 'test',
+							}),
+						],
+						networks: [DEFAULT_NETWORK],
+					},
+					false,
+				),
+			);
+			await testfs({
+				[pathOnRoot(
+					path.join(updateLock.lockPath(123, 'test'), 'updates.lock'),
+				)]: testfs.file({ uid: 0 }),
+				[pathOnRoot(
+					path.join(updateLock.lockPath(123, 'test'), 'resin-updates.lock'),
+				)]: testfs.file({ uid: 0 }),
+			}).enable();
+
+			const errMsg = 'Lockfile exists for { appId: 123, service: test }';
+			await expect(deviceState.shutdown({ reboot: true }))
+				.to.eventually.be.rejectedWith(errMsg)
+				.and.be.an.instanceOf(UpdatesLockedError);
+			await expect(deviceState.shutdown())
+				.to.eventually.be.rejectedWith(errMsg)
+				.and.be.an.instanceOf(UpdatesLockedError);
+
+			await testfs.restore();
+			(applicationManager.getCurrentApps as sinon.SinonStub).restore();
+		});
 
-		(updateLock.abortIfHUPInProgress as sinon.SinonStub).restore();
+		it('prevents reboot or shutdown while update lock module is under process lock', async () => {
+			// Set up a Supervisor service lock
+			sinon.stub(applicationManager, 'getCurrentApps').resolves(
+				createApps(
+					{
+						services: [
+							await createService({
+								running: true,
+								appId: 123,
+								serviceName: 'test',
+							}),
+						],
+						networks: [DEFAULT_NETWORK],
+					},
+					false,
+				),
+			);
+			await testfs({
+				[pathOnRoot(updateLock.lockPath(123, 'test'))]: {
+					'ignored.lock': testfs.file({ uid: updateLock.LOCKFILE_UID }),
+				},
+			}).enable();
+
+			let shouldResolve = false;
+			// Generate a function which doesn't resolve until specified
+			const waitToResolve = async () => {
+				while (!shouldResolve) {
+					await setTimeout(100);
+				}
+				return true;
+			};
+
+			// Take lock and execute the function, which will keep the process lock
+			// until the function is resolved
+			const lockPromise = updateLock.lock(123, { force: false }, waitToResolve);
+			await setTimeout(500);
+			expect(inspect(lockPromise)).to.include('pending');
+
+			// While the Supervisor lock is taken in another context, reboot & shutdown
+			// should wait until locks are freed, so the shutdown promise is pending
```

Comment:
This is not correct, if the supervisor already has the lock that means the reboot/shutdown can happen without waiting. If we wait to release the lock, that means the lock can be taken by the user app before we have a chance to reboot.

However, it is true that each `lock(appId)` call should be executed entirely before another can be processed, which is why the tests passes